share/eds/p2p/server.go

```diff
@@ -0,0 +1,161 @@
+func NewServer(host host.Host, store *eds.Store, protocolSuffix string) *Server {
+	return &Server{
+		host:       host,
+		store:      store,
+		protocolID: protocolID(protocolSuffix),
+	}
+}
+
+func (s *Server) Start(context.Context) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	s.host.SetStreamHandler(s.protocolID, s.handleStream)
+	return nil
+}
+
+func (s *Server) Stop(context.Context) error {
+	defer s.cancel()
+	s.host.RemoveStreamHandler(s.protocolID)
+	return nil
+}
+
+func (s *Server) handleStream(stream network.Stream) {
+	log.Debug("server: handling eds request")
+
+	// read request from stream to get the dataHash for store lookup
+	req, err := s.readRequest(stream)
+	if err != nil {
+		log.Errorw("server: reading request from stream", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	// ensure the requested dataHash is a valid root
+	hash := share.DataHash(req.Hash)
+	err = hash.Validate()
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(s.ctx, readDeadline)
+	defer cancel()
+	status := p2p_pb.Status_OK
+	// determine whether the EDS is available in our store
+	edsReader, err := s.store.GetCAR(ctx, hash)
+	if err != nil {
+		status = p2p_pb.Status_NOT_FOUND
+	} else {
+		defer edsReader.Close()
+	}
+
+	// inform the client of our status
+	err = s.writeStatus(status, stream)
+	if err != nil {
+		log.Errorw("server: writing status to stream", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+	// if we cannot serve the EDS, we are already done
+	if status != p2p_pb.Status_OK {
+		stream.Close()
+		return
+	}
+
+	// start streaming the ODS to the client
+	err = s.writeODS(edsReader, stream)
+	if err != nil {
+		log.Errorw("server: writing ods to stream", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	err = stream.Close()
+	if err != nil {
+		log.Errorw("server: closing stream", "err", err)
+	}
+}
+
+func (s *Server) readRequest(stream network.Stream) (*p2p_pb.EDSRequest, error) {
+	err := stream.SetReadDeadline(time.Now().Add(readDeadline))
+	if err != nil {
+		log.Debug(err)
+	}
+
+	req := new(p2p_pb.EDSRequest)
+	_, err = serde.Read(stream, req)
+	if err != nil {
+		return nil, err
+	}
+	err = stream.CloseRead()
+	if err != nil {
+		log.Error(err)
+	}
+
+	return req, nil
+}
+
+func (s *Server) writeStatus(status p2p_pb.Status, stream network.Stream) error {
+	err := stream.SetWriteDeadline(time.Now().Add(writeDeadline))
+	if err != nil {
+		log.Debug(err)
+	}
+
+	resp := &p2p_pb.EDSResponse{Status: status}
+	_, err = serde.Write(stream, resp)
+	return err
+}
+
+func (s *Server) writeODS(edsReader io.ReadCloser, stream network.Stream) error {
+	err := stream.SetWriteDeadline(time.Now().Add(writeDeadline))
+	if err != nil {
+		log.Debug(err)
+	}
+
+	odsReader, err := eds.ODSReader(edsReader)
+	if err != nil {
+		return fmt.Errorf("creating ODS reader: %w", err)
+	}
+	_, err = io.Copy(stream, odsReader)
```

Comment:
Copy internally creates a buffer with default size. Just realized that we should have more control over it and use CopyBuffer instead, which allows setting a custom buffer and size. We can go with 32kb as default, which we should allow to configure

Reply:
Agree that we should have a const, adding now

Disagree it should be configurable for the same reason the timeouts should not be configurable.

Reply:
All of this will be configurable eventually 😉

Reply:
Hope not.